### ISSUE TITLE
Populate on virtuals is not working.

### ISSUE THREAD

**Issue by sulaysumaria** [Automattic/mongoose#11047]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
Populate on virtuals is not working.

**If the current behavior is a bug, please provide the steps to reproduce.**

```
// Schema:
const schema = new Schema(
  {
    id: { type: String, unique: true },

    mediaId: { type: String, default: '' },
  },
  { timestamps: true, versionKey: false, selectPopulatedPaths: false, toJSON: { virtuals: true } },
);

schema.virtual('media', { ref: 'files', localField: 'mediaId', foreignField: 'id', justOne: true });

// Usage:
await MyModel.findOne({ id }).populate([{ path: 'media' }]);

// Above works perfectly fine with v6.0.13, but does not work with v6.0.14
```

**What is the expected behavior?**
It should populate the media field.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
NodeJS - 16.13.0
Mongoose - 6.0.14
MongoDB - 5.0.4

I am not using the default `_id` for populating but a separate id field (String). Can that be an issue?

**Comment by Kamikadze4GAME** [comment]
Hello @sulaysumaria

Please provide files schema

**Comment by sulaysumaria** [comment]
@Kamikadze4GAME , All my schemas are having similar structure. Here is my Files schema:

```
const fileSchema = new Schema(
  {
    id: { type: String, unique: true },

    fileName: { type: String, default: '' }
  },
  { timestamps: true, versionKey: false, selectPopulatedPaths: false, toJSON: { virtuals: true } },
);
```

**Comment by IslandRhythms** [comment]
Works fine for me

```javascript
const mongoose = require('mongoose');
const {Schema} = mongoose;



const schema = new Schema(
    {
      id: { type: String, unique: true },
  
      mediaId: { type: String, default: '' },
    },
    { timestamps: true, versionKey: false, selectPopulatedPaths: false, toJSON: { virtuals: true }, toObject: {virtuals: true} },
  );
  
  const fileSchema = new Schema(
    {
      id: { type: String, unique: true },
  
      fileName: { type: String, default: '' }
    },
    { timestamps: true, versionKey: false, selectPopulatedPaths: false, toJSON: { virtuals: true }, toObject: {virtuals: true} },
  );


  schema.virtual('media', { ref: 'files', localField: 'mediaId', foreignField: 'id', justOne: true });

  const files = mongoose.model('files', fileSchema);

  const MyModel = mongoose.model('Test', schema);

  async function run() {
    await mongoose.connect('mongodb://localhost:27017/', {useNewUrlParser: true,
    useUnifiedTopology: true,});
    await mongoose.connection.dropDatabase();
    const entry = await MyModel.create({
        id: '123456789012',
        mediaId: '098765432109'
    });
    await files.create({
        id: '098765432109',
        fileName: 'Test'
    });

    const test = await MyModel.findOne({ id: entry.id }).populate({ path: 'media' });
    console.log(test);
  }

  run();
```

**Comment by sulaysumaria** [comment]
@IslandRhythms , I see you added the `toObject: { virtuals: true }`. I added that option, It works for the script, but it still does not work for my project.

If you remove the `toObject` from options, then it does not work.

If I add that option or not, in both case, the queries for finding in file collection is triggered.

Mongoose debug logs:
With `toObject` option.
```
Mongoose: Test.findOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k' }, { projection: { _id: 0, id: 1, mediaId: 1, media: 1 } })
Mongoose: files.find({ id: { '$in': [ 'hR5UNSvwxCO4h8i3wAHqJkVP' ], [Symbol(mongoose#trustedSymbol)]: true }}, { skip: undefined, limit: undefined, perDocumentLimit: undefined, projection: { id: 1, fileName: 1 }})
result:
{
  id: 'JmGvkmbTyPRIdzjny5r5i08k',
  mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP',
  media: { id: 'hR5UNSvwxCO4h8i3wAHqJkVP', fileName: 'Test' }
}
```

Without `toObject` option.
```
Mongoose: Test.findOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k' }, { projection: { _id: 0, id: 1, mediaId: 1, media: 1 } })
Mongoose: files.find({ id: { '$in': [ 'hR5UNSvwxCO4h8i3wAHqJkVP' ], [Symbol(mongoose#trustedSymbol)]: true }}, { skip: undefined, limit: undefined, perDocumentLimit: undefined, projection: { id: 1, fileName: 1 }})
result:
{ id: 'JmGvkmbTyPRIdzjny5r5i08k', mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP' }
```

So, the query is still being triggered, but the output does not contain the `media` field.

Updated script:
```
const mongoose = require('mongoose');
const { Schema } = mongoose;

mongoose.set('debug', true);

const schema = new Schema(
  {
    id: { type: String, unique: true },
    mediaId: { type: String, default: '' },
  },
  { timestamps: true, versionKey: false, selectPopulatedPaths: false, toJSON: { virtuals: true } },
);

const fileSchema = new Schema(
  {
    id: { type: String, unique: true },
    fileName: { type: String, default: '' },
  },
  { timestamps: true, versionKey: false, selectPopulatedPaths: false, toJSON: { virtuals: true } },
);

schema.virtual('media', { ref: 'files', localField: 'mediaId', foreignField: 'id', justOne: true });

const File = mongoose.model('files', fileSchema, 'files');
const MyModel = mongoose.model('Test', schema, 'Test');

async function init() {
  console.log(mongoose.version);

  await mongoose.connect('mongodb://localhost:27017/bugtest', { useNewUrlParser: true, useUnifiedTopology: true, keepAlive: 1 });
  await mongoose.connection.dropDatabase();

  await File.create({ id: 'hR5UNSvwxCO4h8i3wAHqJkVP', fileName: 'Test' });
  await MyModel.create({ id: 'JmGvkmbTyPRIdzjny5r5i08k', mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP' });

  const test = await MyModel.findOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k' })
    .select('-_id id mediaId media')
    .populate([{ path: 'media', select: '-_id id fileName' }]);

  console.log('result:');
  console.log(test);
}

init();

```

**Comment by sulaysumaria** [comment]
@IslandRhythms , I noticed that the way I connect to MongoDB in my project is different. I have multiple database connections so I use `myMongoose.createConnection` instead of `mongoose.connect`.

I updated the script with that connection approach and noticed that populate does not work.

Also to note that below script works fine on `6.0.13` but not on `6.0.14`. Also tried on `6.0.15` and still does not work.

Another thing I noted is the createIndex is also not being called. So indexes are not being created.

Outputs:

```
6.0.13
Primary database connected.
Mongoose: files.deleteMany({}, {})
Mongoose: files.createIndex({ id: 1 }, { background: true })
Mongoose: Test.createIndex({ id: 1 }, { background: true })
Mongoose: Test.deleteMany({}, {})
Mongoose: files.createIndex({ fileName: 1 }, { background: true })
Mongoose: Test.createIndex({ mediaId: 1 }, { background: true })
Mongoose: files.insertOne({ id: 'hR5UNSvwxCO4h8i3wAHqJkVP', fileName: 'Test', _id: new ObjectId("61aef18019e2b4f976e4a709"), createdAt: new Date("Tue, 07 Dec 2021 05:30:40 GMT"), updatedAt: new Date("Tue, 07 Dec 2021 05:30:40 GMT")}, { session: null })
Mongoose: Test.insertOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k', mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP', _id: new ObjectId("61aef18019e2b4f976e4a70b"), createdAt: new Date("Tue, 07 Dec 2021 05:30:40 GMT"), updatedAt: new Date("Tue, 07 Dec 2021 05:30:40 GMT")}, { session: null })
Mongoose: Test.findOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k' }, { projection: { _id: 0, id: 1, mediaId: 1, media: 1 } })
Mongoose: files.find({ id: { '$in': [ 'hR5UNSvwxCO4h8i3wAHqJkVP' ], [Symbol(mongoose#trustedSymbol)]: true }}, { skip: undefined, limit: undefined, perDocumentLimit: undefined, projection: { id: 1, fileName: 1 }})
result:
{
  id: 'JmGvkmbTyPRIdzjny5r5i08k',
  mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP',
  media: { id: 'hR5UNSvwxCO4h8i3wAHqJkVP', fileName: 'Test' }
}
```

```
6.0.15
Primary database connected.
Mongoose: files.deleteMany({}, {})
Mongoose: Test.deleteMany({}, {})
Mongoose: files.insertOne({ id: 'hR5UNSvwxCO4h8i3wAHqJkVP', fileName: 'Test', _id: new ObjectId("61aef1aee8affaa3b8c9e8e9"), createdAt: new Date("Tue, 07 Dec 2021 05:31:26 GMT"), updatedAt: new Date("Tue, 07 Dec 2021 05:31:26 GMT"), __v: 0}, { session: null })
Mongoose: Test.insertOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k', mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP', _id: new ObjectId("61aef1aee8affaa3b8c9e8eb"), createdAt: new Date("Tue, 07 Dec 2021 05:31:26 GMT"), updatedAt: new Date("Tue, 07 Dec 2021 05:31:26 GMT"), __v: 0}, { session: null })
Mongoose: Test.findOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k' }, { projection: { _id: 0, id: 1, mediaId: 1, media: 1 } })
Mongoose: files.find({ id: { '$in': [ 'hR5UNSvwxCO4h8i3wAHqJkVP' ], [Symbol(mongoose#trustedSymbol)]: true }}, { skip: undefined, limit: undefined, perDocumentLimit: undefined, projection: { id: 1, fileName: 1 }})
result:
{ id: 'JmGvkmbTyPRIdzjny5r5i08k', mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP' }
```

Updated script:
```
const mongoose = require('mongoose');
const { Schema } = mongoose;

const primaryDatabaseMongoose = new mongoose.Mongoose();
primaryDatabaseMongoose.set('debug', true);
const primaryDatabaseConnection = primaryDatabaseMongoose.createConnection('mongodb://localhost:27017/bugtest', {
  keepAlive: 1,
  useNewUrlParser: true,
  useUnifiedTopology: true,
});

primaryDatabaseConnection.on('open', () => console.log(`Primary database connected.`));
primaryDatabaseConnection.on('error', (e) => console.log(`Primary database error. ${e}`));

const schema = new Schema(
  {
    id: { type: String, default: '' },
    mediaId: { type: String, default: '' },
  },
  { timestamps: true, versionKey: false, selectPopulatedPaths: false, toJSON: { virtuals: true }, toObject: { virtuals: true } },
);

schema.virtual('media', { ref: 'files', localField: 'mediaId', foreignField: 'id', justOne: true });
schema.index({ id: 1 });
schema.index({ mediaId: 1 });

const fileSchema = new Schema(
  {
    id: { type: String, default: '' },
    fileName: { type: String, default: '' },
  },
  { timestamps: true, versionKey: false, selectPopulatedPaths: false, toJSON: { virtuals: true }, toObject: { virtuals: true } },
);

fileSchema.index({ id: 1 });
fileSchema.index({ fileName: 1 });

const File = primaryDatabaseConnection.model('files', fileSchema, 'files');
const MyModel = primaryDatabaseConnection.model('Test', schema, 'Test');

async function init() {
  console.log(mongoose.version);

  await File.deleteMany();
  await MyModel.deleteMany();

  await File.create({ id: 'hR5UNSvwxCO4h8i3wAHqJkVP', fileName: 'Test' });
  await MyModel.create({ id: 'JmGvkmbTyPRIdzjny5r5i08k', mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP' });

  const test = await MyModel.findOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k' })
    .select('-_id id mediaId media')
    .populate([{ path: 'media', select: '-_id id fileName' }]);

  console.log('result:');
  console.log(test);
}

init();

```

**Comment by sulaysumaria** [comment]
Reverting [this commit](https://github.com/Automattic/mongoose/commit/f37577d632184351aa635db1c0425da7e9667f11) fixes the issue. (tried on my local machine)

@Kamikadze4GAME @IslandRhythms

**Comment by IslandRhythms** [comment]
You need `toObject` to get it to show up

> Keep in mind that virtuals are not included in toJSON() and toObject() output by default. If you want populate virtuals to show up when using functions like Express' res.json() function or console.log(), set the virtuals: true option on your schema's toJSON and toObject() options.

https://mongoosejs.com/docs/populate.html#populate-virtuals

**Comment by sulaysumaria** [comment]
@IslandRhythms , I tried with `toObject`. It still does not show up. Check the Updated github gist I shared above.

**Comment by IslandRhythms** [comment]
@sulaysumaria Upon further testing, it still works without it.

```
const mongoose = require('mongoose');
const {Schema} = mongoose;



const schema = new Schema(
    {
      id: { type: String, unique: true },
  
      mediaId: { type: String, default: '' },
    },
    { timestamps: true, versionKey: false, selectPopulatedPaths: false, toJSON: { virtuals: true }, },
  );
  
  const fileSchema = new Schema(
    {
      id: { type: String, unique: true },
  
      fileName: { type: String, default: '' }
    },
    { timestamps: true, versionKey: false, selectPopulatedPaths: false, toJSON: { virtuals: true },  },
  );


  schema.virtual('media', { ref: 'files', localField: 'mediaId', foreignField: 'id', justOne: true });

  const files = mongoose.model('files', fileSchema);

  const MyModel = mongoose.model('Test', schema);

  async function run() {
    await mongoose.connect('mongodb://localhost:27017/', {useNewUrlParser: true,
    useUnifiedTopology: true,});
    await mongoose.connection.dropDatabase();
    const entry = await MyModel.create({
        id: '123456789012',
        mediaId: '098765432109'
    });
    await files.create({
        id: '098765432109',
        fileName: 'Test'
    });

    const test = await MyModel.findOne({ id: entry.id }).populate({ path: 'media' });
    console.log(test.media);
  }

  run();
  ```
  
  output:
  
  ```
  {
  _id: new ObjectId("61af8161f72bf0358537169e"),
  id: '098765432109',
  fileName: 'Test',
  createdAt: 2021-12-07T15:44:33.425Z,
  updatedAt: 2021-12-07T15:44:33.425Z
}
  ```

**Comment by sulaysumaria** [comment]
> @IslandRhythms , I noticed that the way I connect to MongoDB in my project is different. I have multiple database connections so I use `myMongoose.createConnection` instead of `mongoose.connect`.
> 
> I updated the script with that connection approach and noticed that populate does not work.
> 
> Also to note that below script works fine on `6.0.13` but not on `6.0.14`. Also tried on `6.0.15` and still does not work.
> 
> Another thing I noted is the createIndex is also not being called. So indexes are not being created.
> 
> Outputs:
> 
> ```shell
> 6.0.13
> Primary database connected.
> Mongoose: files.deleteMany({}, {})
> Mongoose: files.createIndex({ id: 1 }, { background: true })
> Mongoose: Test.createIndex({ id: 1 }, { background: true })
> Mongoose: Test.deleteMany({}, {})
> Mongoose: files.createIndex({ fileName: 1 }, { background: true })
> Mongoose: Test.createIndex({ mediaId: 1 }, { background: true })
> Mongoose: files.insertOne({ id: 'hR5UNSvwxCO4h8i3wAHqJkVP', fileName: 'Test', _id: new ObjectId("61aef18019e2b4f976e4a709"), createdAt: new Date("Tue, 07 Dec 2021 05:30:40 GMT"), updatedAt: new Date("Tue, 07 Dec 2021 05:30:40 GMT")}, { session: null })
> Mongoose: Test.insertOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k', mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP', _id: new ObjectId("61aef18019e2b4f976e4a70b"), createdAt: new Date("Tue, 07 Dec 2021 05:30:40 GMT"), updatedAt: new Date("Tue, 07 Dec 2021 05:30:40 GMT")}, { session: null })
> Mongoose: Test.findOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k' }, { projection: { _id: 0, id: 1, mediaId: 1, media: 1 } })
> Mongoose: files.find({ id: { '$in': [ 'hR5UNSvwxCO4h8i3wAHqJkVP' ], [Symbol(mongoose#trustedSymbol)]: true }}, { skip: undefined, limit: undefined, perDocumentLimit: undefined, projection: { id: 1, fileName: 1 }})
> result:
> {
>   id: 'JmGvkmbTyPRIdzjny5r5i08k',
>   mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP',
>   media: { id: 'hR5UNSvwxCO4h8i3wAHqJkVP', fileName: 'Test' }
> }
> ```
> 
> ```shell
> 6.0.15
> Primary database connected.
> Mongoose: files.deleteMany({}, {})
> Mongoose: Test.deleteMany({}, {})
> Mongoose: files.insertOne({ id: 'hR5UNSvwxCO4h8i3wAHqJkVP', fileName: 'Test', _id: new ObjectId("61aef1aee8affaa3b8c9e8e9"), createdAt: new Date("Tue, 07 Dec 2021 05:31:26 GMT"), updatedAt: new Date("Tue, 07 Dec 2021 05:31:26 GMT"), __v: 0}, { session: null })
> Mongoose: Test.insertOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k', mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP', _id: new ObjectId("61aef1aee8affaa3b8c9e8eb"), createdAt: new Date("Tue, 07 Dec 2021 05:31:26 GMT"), updatedAt: new Date("Tue, 07 Dec 2021 05:31:26 GMT"), __v: 0}, { session: null })
> Mongoose: Test.findOne({ id: 'JmGvkmbTyPRIdzjny5r5i08k' }, { projection: { _id: 0, id: 1, mediaId: 1, media: 1 } })
> Mongoose: files.find({ id: { '$in': [ 'hR5UNSvwxCO4h8i3wAHqJkVP' ], [Symbol(mongoose#trustedSymbol)]: true }}, { skip: undefined, limit: undefined, perDocumentLimit: undefined, projection: { id: 1, fileName: 1 }})
> result:
> { id: 'JmGvkmbTyPRIdzjny5r5i08k', mediaId: 'hR5UNSvwxCO4h8i3wAHqJkVP' }
> ```
> 
> Updated script: https://gist.github.com/sulaysumaria/3cf18ea9a6485c4b4c75bcd2e3e50cf2

@IslandRhythms , If you can take a look at this. Looks like it is no longer related to populate but how multiple database connections are handled.

**Comment by IslandRhythms** [comment]
`  console.log(test.media)` add this to the bottom of your gist, it is still there

**Comment by IslandRhythms** [comment]
@vkarpov15 I had to add `console.log(test.media)` to get the populate to show up but since `toObject: {virtuals: true}` is set shouldn't it already be popping up on its own?

**Comment by vkarpov15** [comment]
@sulaysumaria you're right that [this commit](https://github.com/Automattic/mongoose/commit/f37577d632184351aa635db1c0425da7e9667f11) caused this issue in a subtle and indirect way. The issue isn't with populate virtuals, they work fine. The issue is that the virtual wasn't showing up in `toObject()` output because the schema wasn't getting cloned properly. Fix will be in 6.1.5 :+1:
